### PR TITLE
authn: extend authenticator.Token to support audience validation

### PR DESCRIPTION
**Pull request by mikedanese**
```release-note
TokenReview now supports audience validation of tokens with audiences other than the kube-apiserver. 
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cache_striped.go [hunk 1]
```diff
@@ -24,36 +24,36 @@ import (
 // split cache lookups across N striped caches
 type stripedCache struct {
 	stripeCount uint32
-	keyFunc     func(string) uint32
+	hashFunc    func(string) uint32
 	caches      []cache
 }
 
-type keyFunc func(string) uint32
+type hashFunc func(string) uint32
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cache_striped.go]: if we're adding audiences, they have to be part of the cache input

File: staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go [hunk 2]
```diff
@@ -59,24 +60,29 @@ func newWithClock(authenticator authenticator.Token, successTTL, failureTTL time
...
 	case !ok && a.failureTTL > 0:
-		a.cache.set(token, &cacheRecord{user: user, ok: ok, err: err}, a.failureTTL)
+		a.cache.set(key, &cacheRecord{user: user, ok: ok, err: err}, a.failureTTL)
 	}
 
 	return user, ok, err
 }
+
+func keyFunc(auds []string, token string) string {
+	return fmt.Sprintf("%#v|%q", auds, token)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator.go]: ah, missed this

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go [hunk 3]
```diff
@@ -50,13 +50,18 @@ func init() {
...
+func WithAuthentication(handler http.Handler, mapper genericapirequest.RequestContextMapper, auth authenticator.Request, failed http.Handler, apiAuds []string) http.Handler {
 	if auth == nil {
 		glog.Warningf("Authentication is disabled")
 		return handler
 	}
 	return genericapirequest.WithRequestContext(
 		http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			if len(apiAuds) > 0 {
+				if ctx, ok := mapper.Get(req); ok {
+					mapper.Update(req, authenticator.WithAudiences(ctx, apiAuds))
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: is this what we should be updating, or `req.WithContext(authenticator.WithAudiences(req.Context(), apiAuds))`?

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: separately, I wonder if we even need request context mapper now that we can set context directly on the request object, but that's an issue for another PR

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 4]
```diff
@@ -44,11 +45,11 @@ type Password interface {
 }
 
 // TokenFunc is a function that implements the Token interface.
-type TokenFunc func(token string) (user.Info, bool, error)
+type TokenFunc func(auds []string, token string) (user.Info, bool, error)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: need to spec the following:
* what does nil/empty auds mean?
* what should a token handler that is not audience aware do with auds?

File: staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go (hunk 3 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: Ugh, that is not at all confusing.

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: I'm putting together a PR to excise the requestContextMapper from our codebase... shouldn't be necessary any longer

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go]: opened https://github.com/kubernetes/kubernetes/pull/62810

**Comment by ericchiang** [comment]
This is to let TokenReview start handling audiences, right?

**Comment by mikedanese** [comment]
@ericchiang that's the goal.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 4 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: I added an explanation of the semantics on the authenticator.Token interface

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go [hunk 5]
```diff
@@ -523,7 +523,7 @@ func (r *claimResolver) resolve(endpoint endpoint, allClaims claims) error {
 	return nil
 }
 
-func (a *Authenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+func (a *Authenticator) AuthenticateToken(aud []string, token string) (user.Info, bool, error) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/oidc/oidc.go]: our oidc authenticator already checks audience based on the `--oidc-client-id`. should `--oidc-client-id` be added to the kube-apiserver's `AuthenticationInfo.APIAudiences`? should this method short-circuit if `aud` is populated and does not contain ClientID?

File: pkg/serviceaccount/jwt.go [hunk 6]
```diff
@@ -139,7 +139,7 @@ type Validator interface {
 	NewPrivateClaims() interface{}
 }
 
-func (j *jwtTokenAuthenticator) AuthenticateToken(tokenData string) (user.Info, bool, error) {
+func (j *jwtTokenAuthenticator) AuthenticateToken(auds []string, tokenData string) (user.Info, bool, error) {
```
Comment (liggitt) [inline pkg/serviceaccount/jwt.go]: I expected this to make use of auds if provided, prior to `j.validator.Validate`

```
if len(auds) > 0 && len(public.Audience) > 0 && !contains(auds, public.Audience) {
  return nil, false, 
}
```

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go [hunk 7]
```diff
@@ -22,11 +22,15 @@ import (
...
-// Token checks a string value against a backing authentication store and returns
-// information about the current user and true if successful, false if not successful,
-// or an error if the token could not be checked.
+// Token checks a token string and expected audiences against a backing
+// authentication store and returns information about the current user and true
+// if successful, false if not successful, or an error if the token could not
+// be checked. A token handler that is not audience aware, may choose to ignore
+// the audience argument. A nil or empty audience argument means that no
+// audiences were specified. In that case, an audience aware token
+// authenticator should reject the token.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: > A token handler that is not audience aware may choose to ignore the audience argument.

That would mean many of the existing token authenticators would return true for any audience. Ideally, those would only return true when we were checking token validity for the apiserver, not arbitrary clients.

> A nil or empty audience argument means that no audiences were specified.

Under what circumstances would we not pass audience in to the authenticator? The two entry points I know of are the authentication filter and the token review handler. I'd expect both of those places to specify at least one audience. When satisfying token review requests, aren't we taking empty audience to mean "valid against the API server"?

I was hoping we could get to something like this:

|                                                | auds=["","https://kubernetes.default.svc","my-apiserver-oidc-client"] | auds=["myapp"] |
|------------------------------------------------|-----------------------------------------------------------------------|----------------|
| old serviceaccount token                       | ok                                                                    |                |
| new serviceaccount token aud=https://kubernetes.default.svc         | ok                                                                    |                |
| new serviceaccount token aud=myapp             |                                                                       | ok             |
| bootstrap token                                | ok                                                                    |                |
| oidc with client-id=my-apiserver-oidc-client   | ok                                                                    |                |
| token-file token                               | ok                                                                    |                |
| webhook audience unaware                       | ok                                                                    |                |
| webhook audience aware, apiserver token, aud=? | ok (how do we know what to include in auds to make the remote endpoint pass?)                                                                    |                |
| webhook audience aware, myapp token            |                                                                       | ok             |

Notes:
* `auds` is always set when calling `AuthenticateToken`
* when authenticating tokens used against the apiserver, `""` could be included in the list of acceptable audiences, giving token authenticators that are audience-unaware a signal they can authenticate.
* the OIDC token authenticator is already audience-aware. we might need to include the `--oidc-client-id` parameter value in the set of audiences the apiserver identifies itself as
* it would be good to find a way to make an audience-unaware webhook authenticators only authenticate tokens for the apiserver, not arbitrary audiences. one possibility is having them echo the accepted audience in the TokenReview status, defaulting to "", and verified to be in the list of submitted audiences by the webhook authenticator. when checking for the apiserver, no audience would be echoed in the status, defaulting to "", which would be in the list of acceptable candidates. when checking for myapp, the defaulted accepted audience of "" in the status would not match, and would prevent the authentication from succeeding.
* similar to the OIDC authenticator, webhook authenticators may already be audience-aware, but we have no way of knowing what audience it might already be verifying inside the tokens sent to it. If we're using an audience-aware webhook authenticator, we need to think about how to include the audience it is verifying in our list of "audiences the apiserver identifies as"

Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: I think we need clarity over what gets passed for `auds` when:
* authenticating a token request to the API server
* handling a TokenReview with `Audiences:[]`
* handling a TokenReview with `Audiences:[foo]`

**Comment by dims** [comment]
/retest

(not sure if this is a flake)

**Comment by dims** [comment]
@mikedanese looks like a legit failure. can you please look?

File: pkg/serviceaccount/jwt.go (hunk 6 above)
Comment (mikedanese) [reply pkg/serviceaccount/jwt.go]: Done.

File: pkg/apis/authentication/types.go [hunk 8]
```diff
@@ -68,6 +74,13 @@ type TokenReviewStatus struct {
 	Authenticated bool
 	// User is the UserInfo associated with the provided token.
 	User UserInfo
+	// Audience is a single audience identifier chosen by the authenticator that
+	// is compatible with both the TokenReview and token. This identifier can be
+	// any identifier in the intersection of the TokenReviewSpec audiences and
+	// the token's audiences. A client of the TokenReview API should ensure that
```
Comment (liggitt) [inline pkg/apis/authentication/types.go]: `A client of the TokenReview API that sets the spec.audiences field should ensure ...` might be clearer than "audience-bound TokenReviews"

describe the meaning of an empty `status.audience` field on an `authenticated: true` response (indicates the token is valid for the audience of the Kubernetes api server)

File: pkg/apis/authentication/types.go [hunk 9]
```diff
@@ -59,6 +59,12 @@ type TokenReview struct {
 type TokenReviewSpec struct {
 	// Token is the opaque bearer token.
 	Token string
+	// Audiences is a list of the identifiers that the resource server presented
+	// with the token identifies as. Audience-aware token authenticators will
+	// verify that the token was intended for at least one of the audiences in
+	// this list. If no audiences are provided, the audience will default to the
+	// audience of the Kubernetes apiserver.
```
Comment (liggitt) [inline pkg/apis/authentication/types.go]: should we be explicit about this defaulting to `audiences: [""]` and the "" audience meaning "the Kubernetes apiserver"?

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go (hunk 7 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go]: it might be worthwhile to hoist this table to the design doc for the audience changes, to illustrate how we plan to keep compatibility with existing auth methods and APIs, while improving audience capabilities for webhook and service account authenticators

File: pkg/serviceaccount/jwt.go [hunk 10]
```diff
@@ -168,14 +169,28 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(auds []string, tokenData strin
 		return nil, false, utilerrors.NewAggregate(errlist)
 	}
 
+	var (
+		aud string
+		ok  bool
+	)
+	if len(public.Audience) != 0 {
```
Comment (liggitt) [inline pkg/serviceaccount/jwt.go]: don't we want to validate even when public.Audience is `""`, so that old service accounts don't pass validation for arbitrary audiences, only `""`? if we agree on the table in https://github.com/kubernetes/kubernetes/pull/62692#discussion_r192304746 as a goal, would be good to add test cases for as many of those as possible

File: pkg/registry/authentication/tokenreview/storage.go [hunk 11]
```diff
@@ -67,20 +71,31 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
...
+		auds = r.apiAudiences
+	}
+
+	fakeReq = fakeReq.WithContext(genericapirequest.WithAudiences(fakeReq.Context(), auds))
+
 	resp, ok, err := r.tokenAuthenticator.AuthenticateRequest(fakeReq)
 	tokenReview.Status.Authenticated = ok
 	if err != nil {
 		tokenReview.Status.Error = err.Error()
 	}
```
Comment (liggitt) [inline pkg/registry/authentication/tokenreview/storage.go]: should we be requiring the response to be non-nil, and for resp.Status.Audience to be one of the audiences in tokenReview.Spec.Audiences here? or are we counting on the authenticator impl to do that for us?

File: plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go [hunk 12]
```diff
@@ -89,7 +90,7 @@ func tokenErrorf(s *api.Secret, format string, i ...interface{}) {
 //
 //     ( token-id ).( token-secret )
 //
-func (t *TokenAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+func (t *TokenAuthenticator) AuthenticateToken(auds []string, token string) (*authenticator.Response, bool, error) {
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/bootstrap/bootstrap.go]: require auds to contain "" here?

File: staging/src/k8s.io/apiserver/pkg/authentication/token/tokenfile/tokenfile.go [hunk 13]
```diff
@@ -88,10 +89,10 @@ func NewCSV(path string) (*TokenAuthenticator, error) {
 	}, nil
 }
 
-func (a *TokenAuthenticator) AuthenticateToken(value string) (user.Info, bool, error) {
+func (a *TokenAuthenticator) AuthenticateToken(auds []string, value string) (*authenticator.Response, bool, error) {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/authentication/token/tokenfile/tokenfile.go]: require auds to contain "" here?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 14]
```diff
@@ -70,11 +71,11 @@ func newWithBackoff(tokenReview authenticationclient.TokenReviewInterface, ttl,
 }
 
 // AuthenticateToken implements the authenticator.Token interface.
-func (w *WebhookTokenAuthenticator) AuthenticateToken(token string) (user.Info, bool, error) {
+func (w *WebhookTokenAuthenticator) AuthenticateToken(auds []string, token string) (*authenticator.Response, bool, error) {
 	r := &authentication.TokenReview{
 		Spec: authentication.TokenReviewSpec{Token: token},
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: pass auds to the webhook in the spec

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 15]
```diff
@@ -105,11 +106,13 @@ func (w *WebhookTokenAuthenticator) AuthenticateToken(token string) (user.Info,
 		}
 	}
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: need to verify r.Status.Audience is one of the audiences in `auds` (we are a token review client here and must do the verification described in the TokenReview API doc)

**Comment by mikedanese** [comment]
@liggitt PTAL

**Comment by jimangel** [comment]
@mikedanese does this need docs for the 1.12 branch?

**Comment by mikedanese** [comment]
@jimangel yes it will.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/helpers.go [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+*/
+
+package authenticator
+
+// APIAudiences is an Audiences that contains only the default audience
+// identifier of the Kubernetes API server.
+var APIAudiences = Audiences{""}
+
+// Audiences is a container for the Audiences of a token.
+type Audiences []string
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/authentication/authenticator/helpers.go]: Should have a helper method like `HasKubeAPIServer` to avoid all the `aud.Has("")`

**Review by enj — COMMENTED** [review]
Spoke with @liggitt on this earlier today.  Figure I would review before I lost the context.

File: staging/src/k8s.io/apiserver/pkg/authentication/authenticator/helpers.go [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package authenticator
+
+// APIAudiences is an Audiences that contains only the default audience
+// identifier of the Kubernetes API server.
+var APIAudiences = Audiences{""}
```
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/authentication/authenticator/helpers.go]: that was my fault. looking at it today with fresh eyes, I think I agree.

**Comment by mikedanese** [comment]
This PR is now being used to stage the feature only.

/hold

File: test/integration/auth/svcaccttoken_test.go [hunk 18]
```diff
@@ -77,7 +77,7 @@ func TestServiceAccountTokenCreate(t *testing.T) {
 	// Start the server
 	masterConfig := framework.NewIntegrationTestMasterConfig()
 	masterConfig.GenericConfig.Authorization.Authorizer = authorizerfactory.NewAlwaysAllowAuthorizer()
-	masterConfig.GenericConfig.Authentication.APIAudiences = aud
+	masterConfig.GenericConfig.Authentication.APIAudiences = append(aud, "")
```
Comment (awly) [inline test/integration/auth/svcaccttoken_test.go]: why an empty audience?

File: pkg/serviceaccount/jwt.go [hunk 19]
```diff
@@ -111,18 +111,20 @@ func (j *jwtTokenGenerator) GenerateToken(claims *jwt.Claims, privateClaims inte
 // JWTTokenAuthenticator authenticates tokens as JWT tokens produced by JWTTokenGenerator
 // Token signatures are verified using each of the given public keys until one works (allowing key rotation)
 // If lookup is true, the service account and secret referenced as claims inside the token are retrieved and verified with the provided ServiceAccountTokenGetter
-func JWTTokenAuthenticator(iss string, keys []interface{}, validator Validator) authenticator.Token {
+func JWTTokenAuthenticator(iss string, keys []interface{}, implicitAuds authenticator.Audiences, validator Validator) authenticator.Token {
```
Comment (awly) [inline pkg/serviceaccount/jwt.go]: document what `implicitAuds` does

File: pkg/serviceaccount/jwt.go [hunk 20]
```diff
@@ -170,14 +172,34 @@ func (j *jwtTokenAuthenticator) AuthenticateToken(ctx context.Context, tokenData
...
+	}
+
+	requestedAudiences, ok := authenticator.AudiencesFrom(ctx)
+	if !ok {
+		// default to apiserver audiences
+		requestedAudiences = j.implicitAuds
+	}
+
+	auds := authenticator.Audiences(tokenAudiences).Intersect(requestedAudiences)
+	if len(auds) == 0 && len(j.implicitAuds) != 0 {
```
Comment (awly) [inline pkg/serviceaccount/jwt.go]: What if implcitAuds is empty but tokenAudiences and requestedAudiences are non-empty non-overlapping sets?

File: test/integration/serviceaccount/service_account_test.go [hunk 21]
```diff
@@ -374,8 +374,8 @@ func startServiceAccountTestServer(t *testing.T) (*clientset.Clientset, restclie
 	})
 	serviceAccountKey, _ := rsa.GenerateKey(rand.Reader, 2048)
 	serviceAccountTokenGetter := serviceaccountcontroller.NewGetterFromClient(rootClientset)
-	serviceAccountTokenAuth := serviceaccount.JWTTokenAuthenticator(serviceaccount.LegacyIssuer, []interface{}{&serviceAccountKey.PublicKey}, serviceaccount.NewLegacyValidator(true, serviceAccountTokenGetter))
-	authenticator := union.New(
```
Comment (awly) [inline test/integration/serviceaccount/service_account_test.go]: revert this change to minimize the diff

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 22]
```diff
@@ -147,6 +149,10 @@ func (s *DelegatingAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 	if s.RemoteKubeConfigFileOptional {
 		optionalKubeConfigSentence = " This is optional. If empty, all token requests are considered to be anonymous and no client CA is looked up in the cluster."
 	}
+	fs.StringSliceVar(&s.APIAudiences, "api-audiences", s.APIAudiences, ""+
+		"Identifiers of the API. Authenticators will validate that tokens used "+
```
Comment (awly) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go]: don't wrap the string

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 23]
```diff
@@ -138,3 +161,7 @@ func (t *tokenReviewClient) Create(tokenReview *authentication.TokenReview) (*au
 	err := t.w.RestClient.Post().Body(tokenReview).Do().Into(result)
 	return result, err
 }
+
+func keyFunc(trs authentication.TokenReviewSpec) string {
```
Comment (awly) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: Is this used anywhere?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook_test.go [hunk 24]
```diff
@@ -378,31 +379,140 @@ func TestWebhookTokenAuthenticator(t *testing.T) {
...
-		}
-		if authenticated != tt.expectedAuthenticated {
-			t.Errorf("case %d: Plugin returned incorrect authentication response. Got %t, expected %t.",
-				i, authenticated, tt.expectedAuthenticated)
-		}
-		if resp != nil && tt.expectedUser != nil && !reflect.DeepEqual(resp.User, tt.expectedUser) {
-			t.Errorf("case %d: Plugin returned incorrect user. Got %#v, expected %#v",
-				i, resp.User, tt.expectedUser)
-		}
+		t.Run(fmt.Sprint(i), func(t *testing.T) {
```
Comment (awly) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook_test.go]: can we give them names?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 25]
```diff
@@ -45,38 +46,47 @@ var _ authenticator.Token = (*WebhookTokenAuthenticator)(nil)
...
 // newWithBackoff allows tests to skip the sleep.
-func newWithBackoff(tokenReview authenticationclient.TokenReviewInterface, initialBackoff time.Duration) (*WebhookTokenAuthenticator, error) {
-	return &WebhookTokenAuthenticator{tokenReview, initialBackoff}, nil
+func newWithBackoff(tokenReview authenticationclient.TokenReviewInterface, initialBackoff time.Duration, implicitAuds authenticator.Audiences) (*WebhookTokenAuthenticator, error) {
+	return &WebhookTokenAuthenticator{tokenReview, initialBackoff, implicitAuds}, nil
 }
 
 // AuthenticateToken implements the authenticator.Token interface.
 func (w *WebhookTokenAuthenticator) AuthenticateToken(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+	reqAuds, checkAuds := authenticator.AudiencesFrom(ctx)
```
Comment (cjcullen) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: It took me a while to make sense of this. Can you add a comment describing that "when no audience is passed in the context passed to the Authenticate call, it means the webhook token authenticator should not validate any audience."

Comment (cjcullen) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: reqAuds & respAuds were confusing to me. Would they make more sense as wantAuds & gotAuds?

File: test/integration/auth/svcaccttoken_test.go (hunk 18 above)
Comment (mikedanese) [reply test/integration/auth/svcaccttoken_test.go]: Removed.

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go (hunk 25 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: Added a comment. https://github.com/kubernetes/kubernetes/issues/69893 has a good outline of what needs to happen here.

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go [hunk 26]
```diff
@@ -147,6 +149,8 @@ func (s *DelegatingAuthenticationOptions) AddFlags(fs *pflag.FlagSet) {
 	if s.RemoteKubeConfigFileOptional {
 		optionalKubeConfigSentence = " This is optional. If empty, all token requests are considered to be anonymous and no client CA is looked up in the cluster."
 	}
+	fs.StringSliceVar(&s.APIAudiences, "api-audiences", s.APIAudiences, "Identifiers of the API. Authenticators will validate that tokens used against the API are bound to at least one of these audiences.")
```
Comment (cjcullen) [inline staging/src/k8s.io/apiserver/pkg/server/options/authentication.go]: "Identifiers of the API" is pretty vague, but I'm struggling to come up with anything much better. Maybe "Identifiers of this apiserver."?

File: staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go [hunk 27]
```diff
@@ -45,38 +44,54 @@ var _ authenticator.Token = (*WebhookTokenAuthenticator)(nil)
...
+func newWithBackoff(tokenReview authenticationclient.TokenReviewInterface, initialBackoff time.Duration, implicitAuds authenticator.Audiences) (*WebhookTokenAuthenticator, error) {
+	return &WebhookTokenAuthenticator{tokenReview, initialBackoff, implicitAuds}, nil
 }
 
 // AuthenticateToken implements the authenticator.Token interface.
 func (w *WebhookTokenAuthenticator) AuthenticateToken(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+	// When no audience is passed in the context passed to the Authenticate call,
+	// the webhook token authenticator should not validate any audiences.
+	// Otherwise, it should check that wantAuds intersects with the webhook's
+	// TokenReviw response audiences. If the webhooks TokenReviw response
```
Comment (cjcullen) [inline staging/src/k8s.io/apiserver/plugin/pkg/authenticator/token/webhook/webhook.go]: TokenReview (in both places)

File: pkg/apis/authentication/types.go [hunk 28]
```diff
@@ -59,6 +59,13 @@ type TokenReview struct {
 type TokenReviewSpec struct {
 	// Token is the opaque bearer token.
 	Token string
+	// Audiences is a list of the identifiers that the resource server presented
+	// with the token identifies as. Audience-aware token authenticators will
+	// verify that the token was intended for at least one of the audiences in
+	// this list. If no audiences are provided, the audience will default to the
```
[the remark this answers is not in the document]
Comment (cjcullen) [reply pkg/apis/authentication/types.go]: I don't think that's the behavior we want. If no audiences are provided in the `TokenReview`, the client is asking "Hey apiserver, is this token valid for you?", so we _do_ need to check that the audience matches this apiserver. (I think the last sentence in this comment is not correct, though)

File: pkg/apis/authentication/types.go [hunk 29]
```diff
@@ -59,6 +59,13 @@ type TokenReview struct {
...
+	// audience of the Kubernetes apiserver. The empty string audience is the
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply pkg/apis/authentication/types.go]: It doesn't any longer.

File: pkg/apis/authentication/types.go [hunk 30]
```diff
@@ -68,6 +75,16 @@ type TokenReviewStatus struct {
...
+	// Audience is a single audience identifier chosen by the authenticator that
+	// is compatible with both the TokenReview and token. This identifier can be
+	// any identifier in the intersection of the TokenReviewSpec audiences and
+	// the token's audiences. A client of the TokenReview API that sets the
+	// spec.audiences field should validate that a compatible audience identifier
+	// is returned in the status.audience field to ensure that the TokenReview
+	// server is audience aware. If a TokenReview returns an empty
+	// status.audience field where status.authenticated is "true", the token is
+	// valid against the audience of the Kubernetes API server.
+	Audience string
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply pkg/apis/authentication/types.go]: Done.

File: pkg/apis/authentication/types.go (hunk 28 above)
[the remark this answers is not in the document]
Comment (mikedanese) [reply pkg/apis/authentication/types.go]: Fixed the last sentence.

File: staging/src/k8s.io/apiserver/pkg/server/options/authentication.go (hunk 26 above)
Comment (mikedanese) [reply staging/src/k8s.io/apiserver/pkg/server/options/authentication.go]: Changed to "Identifiers of this API server. Authenticators will validate that tokens used against this API server are bound to at least one of these audiences."

File: pkg/registry/authentication/tokenreview/storage.go [hunk 31]
```diff
@@ -68,6 +72,13 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 	fakeReq := &http.Request{Header: http.Header{}}
 	fakeReq.Header.Add("Authorization", "Bearer "+tokenReview.Spec.Token)
 
+	auds := tokenReview.Spec.Audiences
+	if len(tokenReview.Spec.Audiences) == 0 {
+		auds = r.apiAudiences
+	}
+
+	fakeReq = fakeReq.WithContext(authenticator.WithAudiences(fakeReq.Context(), auds))
```
Comment (liggitt) [inline pkg/registry/authentication/tokenreview/storage.go]: only do this `if len(auds) > 0`, right?

Comment (liggitt) [reply pkg/registry/authentication/tokenreview/storage.go]: per https://github.com/kubernetes/kubernetes/pull/69916#discussion_r228208287

File: pkg/registry/authentication/tokenreview/storage.go [hunk 32]
```diff
@@ -87,6 +98,7 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 		for k, v := range resp.User.GetExtra() {
 			tokenReview.Status.User.Extra[k] = authentication.ExtraValue(v)
 		}
+		tokenReview.Status.Audiences = resp.Audiences
```
[the remark this answers is not in the document]
Comment (mikedanese) [reply pkg/registry/authentication/tokenreview/storage.go]: Should this cause a 500?

File: pkg/registry/authentication/tokenreview/storage.go [hunk 33]
```diff
@@ -68,14 +76,24 @@ func (r *REST) Create(ctx context.Context, obj runtime.Object, createValidation
 	fakeReq := &http.Request{Header: http.Header{}}
 	fakeReq.Header.Add("Authorization", "Bearer "+tokenReview.Spec.Token)
 
+	auds := tokenReview.Spec.Audiences
+	if len(auds) == 0 {
+		auds = r.apiAudiences
+	}
+	if len(auds) == 0 {
```
Comment (liggitt) [inline pkg/registry/authentication/tokenreview/storage.go]: if len >= 0, right? (and if tests passed with this, add a test that would have caught this)

Comment (mikedanese) [reply pkg/registry/authentication/tokenreview/storage.go]: Nice catch. This area is covered by integration tests which failed. https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/62692/pull-kubernetes-integration/36093/

**Comment by mikedanese** [comment]
Adding labels per https://kubernetes.slack.com/archives/C2C40FMNF/p1542418440380400

**Comment by liggitt** [comment]
can you link the docs PR for the relevant docs update against the dev-1.13 branch?
